exetera/core/field_importers.py

```diff
@@ -0,0 +1,344 @@
+
+        self.field.data.write_part(elements)
+        if self.flag_field is not None:
+            self.flag_field.data.write_part(validity)
+
+
+    def _is_blank(self, value):
+        return (isinstance(value, str) and value.strip() == '') or value == b''
+
+
+    def complete(self):
+        self.field.data.complete()
+        if self.flag_field is not None:
+            self.flag_field.data.complete()
+
+
+class IndexedStringImporter:
+    def __init__(self, session, df, name, timestamp=None):
+        self.field = df.create_indexed_string(name, timestamp, None)
+        self.chunk_accumulated = 0
+        self.field.indices.write_part([0])
+
+    def transform_and_write_part(self, column_inds, column_vals, column_offsets, col_idx, written_row_count):
+        # broadcast accumulated size to current index array
+        index = column_inds[col_idx, :written_row_count + 1] + self.chunk_accumulated
+        self.chunk_accumulated += column_inds[col_idx, written_row_count]
+
+        col_offset = column_offsets[col_idx]
+        values = column_vals[col_offset: col_offset + column_inds[col_idx, written_row_count]]
+        self.write_part(index, values)
+
+    def write_part(self, index, values):
+        if index.dtype != np.int64:
+            raise ValueError(f"'index' must be an ndarray of '{np.int64}'")
+        if values.dtype not in (np.uint8, 'S1'):
+            raise ValueError(f"'values' must be an ndarray of '{np.uint8}' or 'S1'")
+        self.field.indices.write_part(index[1:])
+        self.field.values.write_part(values)
+
+    def complete(self):
+        # self.field.data.complete()
+        self.field.indices.complete()
+        self.field.values.complete()
+
+
+class FixedStringImporter:
+    def __init__(self, session, df, name, strlen, timestamp = None):
+        self.field = df.create_fixed_string(name, strlen, timestamp, None)  
+        self.strlen = strlen
+        self.field_size = strlen        
+
+    def transform_and_write_part(self, column_inds, column_vals, column_offsets,  col_idx, written_row_count):
+        values = np.zeros(written_row_count, dtype='S{}'.format(self.strlen))
+        ops.fixed_string_transform(column_inds, column_vals, column_offsets, col_idx,
+                                   written_row_count, self.strlen, values.data.cast('b'))
+        self.field.data.write_part(values)
+
+    def complete(self):
+        self.field.data.complete()
+
+
+class DateTimeImporter:
+    def __init__(self, session, df, name, create_day_field=False, create_flag_field=False, timestamp=None):
+        self.field = df.create_timestamp(name, timestamp, None)   
+        self.day_field = None
+        if create_day_field:
+            self.day_field = df.create_fixed_string(f"{name}_day", 10, timestamp, None)  
+
+        self.flag_field = None
+        if create_flag_field:
+            self.flag_field = df.create_numeric(f"{name}_set", 'bool', timestamp, None)
+
+    def write_part(self, values):
```

Comment:
Are these write_part methods still required?

Reply:
Yes, it depends on what type of data we get after using transform_and_write_part(), sometimes, we can call self.field.data.write_part() directly. Other times, we still need to do some transformation on data before writing it or adding extra day_fields, flag_field. Otherwise,

Reply:
I think your comment got truncated

Reply:
just updated the comments